src/netblock.py

```diff
@@ -18,72 +18,356 @@
 
+################ Time Buckets
 OneDay = 86400          # seconds per day
-BucketSize = 3600       # seconds per bucket
 
-def time2bucket(time):
-  """ convert seconds to a bucket name string: T00:00 - T23:59 """
-  time = time % OneDay
-  return "T%02d:%02d"%(time/3600, (time/60)%60)
+class TimeBucket():
+  """Tag tests by time, rounded into buckets.
+
+  The bucket size must be multiples of minutes, specified in seconds,
+  and divide into 24 hours.  (Required by other parts of the
+  algorithm).
+  """
+
+  def __init__(self, bucketsize):
+    """Create the class with the specified bucket size.
+    """
+    assert(bucketsize %60 == 0) # Manually tested once
+    assert(OneDay % bucketsize == 0) # Manually tested once
+    self.bucketsize = bucketsize
+
+  def bucket(self,time):
+    """Convert seconds to a bucket name string: T00:00 - T23:59 """
+
+    bucket = ((time / self.bucketsize) * self.bucketsize) % OneDay
+    return "T%02d:%02d"%(bucket/3600, (bucket/60)%60)
+
+  def all(self):
+    """Return a list of all bucket names."""
+
+    return([self.bucket(i) for i in range(0, OneDay, self.bucketsize)])
+
+################ IP address and subnet tools
+from netaddr import IPAddress
+
+def inet_addr(s):
+  """Convert IP address string xxx.xxx.xxx.xxx to int"""
+  return int(IPAddress(s))
+
+def inet_ntoa(i):
+  """Convert int to address string xxx.xxx.xxx.xxx"""
+  return str(IPAddress(i))
+
+class SubNet():
+  """Primitives for creating an manipulating subnet masks"""
+
+  def __init__(self, width, address):
+    """Create a subnet mask from an address at a specified width.
+
+    NB: does not support IPv6
+    """
+    self.width = width
+    self.mask = (2 ** 32) - (2 ** (32 - width))
+    self.prefix = self.mask & address
+    self.last = (2 ** (32 - width)) - 1 + self.prefix
+
+  def match(self, address):
+    """Check if an address matches a subnet
+
+    NB: Match may be expanded inline for panda iterators
+    """
+    return (self.mask & address) == self.prefix
+
+  def invert(self):
+    """Toggle the last bit of the prefix"""
+    prefix = self.prefix ^ (2 ** (32 - self.width))
+    return SubNet(self.width, prefix)
+
+  def str(self):
+    """Display conventional prefix/len notation"""
+    return "%s/%d"%(inet_ntoa(self.prefix), self.width)
+
+  def vstr(self):
+    """Verbose display: prefix/len (netmask) - last"""
+    return "%s/%d (%s) - %s"%(inet_ntoa(self.prefix),
+                              self.width,
+                              inet_ntoa(self.mask),
+                              inet_ntoa(self.last))
+
+################ Calendar queue (aka heap queue)
+import heapq
+def hpush(heap, item):
+  """Queue ordered by item.rank"""
+  heapq.heappush(heap, (item.rank, item))
+
+def hpop(heap):
+  """Dequeue, discard rank"""
+  rank, item = heapq.heappop(heap)
+  return item
 
 ################################################################
-class NetBlock(DataFrame):
+class NetBlock():
```

Comment:
could this inherit from `netaddr.IPNetwork` ?

That might save you a bunch of parsing pain.